src/pk/dh/dh_make_key.c

```diff
@@ -0,0 +1,207 @@
+      return 0;
+   }
+   else if (groupsize <= 192) {
+      return 30;     /* 1536-bit => key size 240-bit */
+   }
+   else if (groupsize <= 256) {
+      return 40;     /* 2048-bit => key size 320-bit */
+   }
+   else if (groupsize <= 384) {
+      return 52;     /* 3072-bit => key size 416-bit */
+   }
+   else if (groupsize <= 512) {
+      return 60;     /* 4096-bit => key size 480-bit */
+   }
+   else if (groupsize <= 768) {
+      return 67;     /* 6144-bit => key size 536-bit */
+   }
+   else if (groupsize <= 1024) {
+      return 77;     /* 8192-bit => key size 616-bit */
+   }
+   else {
+      return 0;
+   }
+}
+
+static int _dh_make_key(prng_state *prng, int wprng, void *prime, void *base, dh_key *key)
+{
+   unsigned char *buf;
+   unsigned long keysize;
+   int err, max_iterations = PK_MAX_RETRIES;
+
+   LTC_ARGCHK(key   != NULL);
+   LTC_ARGCHK(prng  != NULL);
+   LTC_ARGCHK(prime != NULL);
+   LTC_ARGCHK(base  != NULL);
+
+   /* good prng? */
+   if ((err = prng_is_valid(wprng)) != CRYPT_OK) {
+      return err;
+   }
+
+   /* init big numbers */
+   if ((err = mp_init_multi(&key->x, &key->y, &key->base, &key->prime, NULL)) != CRYPT_OK) {
+      return err;
+   }
+
+   /* load the prime and the base */
+   if ((err = mp_copy(base, key->base)) != CRYPT_OK)   { goto freemp; }
+   if ((err = mp_copy(prime, key->prime)) != CRYPT_OK) { goto freemp; }
+
+   keysize = _dh_groupsize_to_keysize(mp_unsigned_bin_size(key->prime));
+   if (keysize == 0) {
+      err = CRYPT_INVALID_KEYSIZE;
+      goto freemp;
+   }
+
+   /* allocate buffer */
+   buf = XMALLOC(keysize);
+   if (buf == NULL) {
+      err = CRYPT_MEM;
+      goto freemp;
+   }
+
+   key->type = PK_PRIVATE;
+   do {
+      /* make up random buf */
+      if (prng_descriptor[wprng].read(buf, keysize, prng) != keysize) {
+         err = CRYPT_ERROR_READPRNG;
+         goto freebuf;
+      }
+      /* load the x value - private key */
+      if ((err = mp_read_unsigned_bin(key->x, buf, keysize)) != CRYPT_OK) {
+         goto freebuf;
+      }
+      /* compute the y value - public key */
+      if ((err = mp_exptmod(key->base, key->x, key->prime, key->y)) != CRYPT_OK) {
+         goto freebuf;
+      }
+      err = dh_check_pubkey(key);
+   } while (err != CRYPT_OK && max_iterations-- > 0);
+
+freebuf:
+   zeromem(buf, keysize);
+   XFREE(buf);
+freemp:
+   if (err != CRYPT_OK) mp_clear_multi(key->x, key->y, key->base, key->prime, NULL);
+   return err;
+}
+
+/**
+  Make a DH key (custom DH group) [private key pair]
+  @param prng       An active PRNG state
+  @param wprng      The index for the PRNG you desire to use
+  @param prime_hex  The prime p (hexadecimal string)
+  @param base_hex   The base g (hexadecimal string)
+  @param key        [out] Where the newly created DH key will be stored
+  @return CRYPT_OK if successful, note: on error all allocated memory will be freed automatically.
+*/
+static int _dh_make_key_ex(prng_state *prng, int wprng, int radix,
```

Comment:
the only place where this in (currently) invoked from is `dh_make_key()`, therefore I'd prefer to discard this function and put the implementation in `dh_make_key()` (without the `radix == 256` part)